Editors/vscode/src/inlayHints.ts

```diff
@@ -0,0 +1,202 @@
+
+        // Set up initial cache
+        this.visibleSourceKitLSPEditors.forEach(editor => this.sourceFiles.set(
+            editor.document.uri.toString(),
+            {
+                document: editor.document,
+                inFlightInlayHints: null,
+                cachedDecorations: null
+            }
+        ));
+
+        this.syncCacheAndRenderHints();
+    }
+
+    private onDidChangeVisibleTextEditors(): void {
+        const newSourceFiles = new Map<string, SourceFile>();
+
+        // Rerender all, even up-to-date editors for simplicity
+        this.visibleSourceKitLSPEditors.forEach(async editor => {
+            const uri = editor.document.uri.toString();
+            const file = this.sourceFiles.get(uri) ?? {
+                document: editor.document,
+                inFlightInlayHints: null,
+                cachedDecorations: null
+            };
+            newSourceFiles.set(uri, file);
+
+            // No text documents changed, so we may try to use the cache
+            if (!file.cachedDecorations) {
+                const hints = await this.fetchHints(file);
+                if (!hints) return;
+                file.cachedDecorations = this.hintsToDecorations(hints);
+            }
+
+            this.renderDecorations(editor, file.cachedDecorations);
+        });
+
+        // Cancel requests for no longer visible (disposed) source files
+        this.sourceFiles.forEach((file, uri) => {
+            if (!newSourceFiles.has(uri)) {
+                file.inFlightInlayHints?.cancel();
+            }
+        });
+
+        this.sourceFiles = newSourceFiles;
+    }
+
+    private onDidChangeTextDocument(event: vscode.TextDocumentChangeEvent): void {
+        if (event.contentChanges.length !== 0 && this.isSourceKitLSPDocument(event.document)) {
+            this.syncCacheAndRenderHints();
+        }
+    }
+
+    private syncCacheAndRenderHints(): void {
+        this.sourceFiles.forEach(async (file, uri) => {
+            const hints = await this.fetchHints(file);
+            if (!hints) return;
+
+            const decorations = this.hintsToDecorations(hints);
+            file.cachedDecorations = decorations;
+
+            this.visibleSourceKitLSPEditors.forEach(editor => {
+                if (editor.document.uri.toString() === uri) {
+                    this.renderDecorations(editor, decorations);
+                }
+            });
+        });
+    }
+
+    private get visibleSourceKitLSPEditors(): vscode.TextEditor[] {
+        return vscode.window.visibleTextEditors.filter(e => this.isSourceKitLSPDocument(e.document));
+    }
+
+    private isSourceKitLSPDocument(document: vscode.TextDocument): boolean {
+        // TODO: Add other SourceKit-LSP languages if/once we forward inlay
+        // hint requests to clangd.
+        return document.languageId === 'swift' && document.uri.scheme === 'file';
```

Comment:
Disclaimer: I am not at all familiar with how VSCode handles open documents and language servers.

Couldn’t the user in theory select some other language server for Swift and use SourceKit-LSP for C-family code only? I’m not sure if the Swift files would then be reported as part of the visible documents, but we might need a more sophisticated way of determining whether we are providing language server functionality for that file in the future.

This is fine for now though.

Reply:
Yes that's true, it's more of a heuristic currently. Once LSP includes inlay hints, it won't be needed anymore anyway, since VSCode would internally handle inlay hint requests and make sure they get sent to the correct language server.